Make mail links work through oauth

xmo-odoo opened the pull request:
### note: not sure whether that's supposed to be fixed in master or in an older version, creating this in master for discussion anyway

Tasks 2234813, 2221168, 2230454 

The "final" URL of the oauth flow is generated by the server, which means it can't access the [fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier), which means while the fragment is kept from `/web#action=xxx` to `/web/login` it can't be merged into the "terminal" URL, and thus gets lost when we go through the authorization server. It works fine for regular login because that's a redirection and fragments get conserved through redirection, but the oauth flow starts through a *link*, which will not keep whatever fragment the current page has (plus we can't necessarily assume the oauth server is not a fragment-using weirdo anyway, I guess).

Anyway this implements a relatively simple fix specifically for chatter notification mails which provide a link to the corresponding object: this link goes through mail's `/mail/view` indirection which uses query string and derefs that to whatever action description is necessary. It's what used for e.g. the
"view task" link in notification emails.

The issue with it, is that it ends up deref'ing to a fragment-based URL, which has the issue notified above. Instead, bounce to /web with a `redirect` query string parameter of that fragment-based URL: [`auth_oauth` will use that parameter as its final URL instead of generating one from scratch](https://github.com/odoo/odoo/blob/de590816d8d06730d99499db9a068f8d0e2cd3f9/addons/auth_oauth/controllers/main.py#L73), and if the user is already logged in `/web` will just redirect to the "final" `/web` invocation, which is one more hop but shouldn't be a significant cost.

Then fix /web so it doesn't lose the `redirect` parameter when redirecting non-logged users to the login page: as noted above `/web` will use that parameter if the user is already logged, and `/web/login` will *also* use it on login (including auth_oauth), it would just go missing when bouncing from one to the other. Which doesn't seem like a security issue though maybe @mart-e will object.

tde-banana-odoo commented:
:postal_horn:

xmo-odoo commented:
Of note: this is still very unlikely to work for reasons independent of the oauth issue, I'm guessing because caf900e89ea292c56ed773232792a2aab19de8d5 converted the endpoint from *none* to *public*.

This means if the target object is a portal mixin (which more or less all relevant objects are) we're going to call `PortalMixin.get_access_action` with a public user which is `share`, which will return an act_url: https://github.com/odoo/odoo/blob/4b38cc6590abcb58a0ca102a06063eb3db7ac1f4/addons/portal/models/portal_mixin.py#L100-L120

Since that's a non-public URL, this will redirect to messaging instead of the actual object: https://github.com/odoo/odoo/blob/536175dd9f6fff1b83ae9d496ecf8bd48dfd2bdb/addons/mail/controllers/main.py#L102-L104

tde-banana-odoo commented:
Do you mean it's currently broken ?

xmo-odoo commented:
ping @tde-banana-odoo after some more testing, `mail_action_view` / `_redirect_to_record` seems to be pretty broken for reasons unrelated to the oauth issue as soon as the user going through it is not logged.

xmo-odoo commented:
> Do you mean it's currently broken ?

Yeah I think if the user is not logged and the target is a portal mixin it's broken now, and it used to be broken but differently. I'm not sure why I didn't stumble upon this issue when I tested the flow yesterday though. I *think* I tested on a crm lead which should have the problem.

xmo-odoo commented:
Note: the "broken now" is probably worse than the "broken before", at least for, say, employees: now, the user is always considered "share", meaning they always get a portal URL, meaning they always get redirected to messaging (which incidentally is a bit weird, public or portal users don't have access to messaging do they?)

Before, the user would always be considered internal (because uid=1), so they'd always be redirected to the object in the backend, which would trigger an access error for portal users but would work fine for internal users.

xmo-odoo commented:
@tde-banana-odoo added a test module for the behaviour of /mail/view's flow, and changed a case which I think doesn't make much sense:

* an internal user gets a link to a non-portal resource they *do not* have access to, they are not logged, after logging-in they will be redirected to the resource and end up with an access error
* an internal user gets a link to a portal resource they *do* have access to, they are not logged in, after logging-in they will be redirected to the chatter instead of the resource)

Fixed by updating the behaviour of the endpoint: if the user is not logged-in, and the action is anything other than a public url, redirects the user to the login page then back to /mail/view to re-resolve the thing now as a logged-in user.

xmo-odoo commented:
Rebased and fixed on top of current master (got hit by both the lazy orm and people breaking redirects for portal users), seems to be working again though will likely be more annoying to backport (if backported)

/cc @tde-banana-odoo @nle-odoo

tde-banana-odoo commented on addons/mail/controllers/mail.py:
```diff
@@ -95,16 +96,25 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
+            if (act['type'], act.get('target_type')) != ('ir.actions.act_url', 'public'):
+                return werkzeug.utils.redirect(
+                    '/web/login?' + url_encode({
+                        'redirect': '/mail/view?' + url_encode({
```
Technically I think we also support message_id and access_token (and pid, hash, auth_signup_token and auth_login in portal). Don't remember if everything is useful and all the flows, woud be interesting to dive in it someday, but we could at least support them and see afterwards :D (or maybe we give kwargs as we receive them and basta ?).

tde-banana-odoo commented on addons/test_portal/models.py:
```diff
@@ -0,0 +1,17 @@
+        super()._compute_access_url()
+        self.access_url = '/my/thing2'
+
+class Portal(models.Model):
```
Could you maybe add some basic fields (email, name, partner_id, ...) ? That way I will re-use it soon to test some notification flows with portal (as test_mail does not check portal). If you don't want because "minimal diff", Rochers Suchards will be required.

tde-banana-odoo commented on addons/test_portal/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+"id","name","model_id:id","group_id:id","perm_read","perm_write","perm_create","perm_unlink"
+test_portal_0,xxx,model_test_portal_portal,base.group_system,1,1,1,1
```
``access_test_portal_portal_system,access.test.portal.portal.system`` ? Just to be coherent with naming used in SM team (at least), aka: main model (access for ir.model.access), used for test_portal_portal_model, for group_system. Just some "top-bottom" naming we try to use.

tde-banana-odoo replied:
Those are in guidelines note.

tde-banana-odoo commented on addons/test_portal/tests/test_mail_links.py:
```diff
@@ -0,0 +1,137 @@
+        })
+        while True:
+            r = c.get(url)
+            self.assertLess(r.status_code, 400, f'{url} = {r.status}')
```
Isn't it equivalent to result.ok ?

xmo-odoo replied:
And of course after actually trying it it doesn't work because `r` is a `werkzeug.wrappers.Response`, not a `requests.Response`, and so does not *have* an `ok` property.

tde-banana-odoo commented on addons/test_portal/tests/test_mail_links.py:
```diff
@@ -0,0 +1,137 @@
+            **params
+        })
+        while True:
+            r = c.get(url)
```
reponse ? result ? r alone does not mean much, especially when you have 2 bytes of memory like me.

tde-banana-odoo commented on addons/test_portal/tests/test_mail_links.py:
```diff
@@ -0,0 +1,137 @@
+        PORTAL_URL = '/my/thing?model={model}&res_id={id}{token}'
+        NO_ACCESS = '/web#action=mail.action_discuss'
+        PORTAL_ROOT = '/my' # portal user gets bounced from /web to /my
+        cases = [
```
Not sure to understand your test setup completely as it is a bit obscure... but if you refer to demo / admin users, you should probably use HttpCaseWithUserDemo class that ensures demo user exists (in case it has been deleted or we are running tests without demo data, which we try to support).

xmo-odoo replied:
Added explanation for the test table. It's really not that obscure.

tde-banana-odoo reviewed (commented):
Blork ! Had a look ! Rochers Suchards !

tde-banana-odoo commented:
:wave:

std-odoo commented on addons/test_portal/controllers.py:
```diff
@@ -0,0 +1,8 @@
+
+class Portal(CustomerPortal):
+    @http.route(['/my/thing', '/my/thing2'], type='http', auth="user", website=True)
+    def thing(self, **_):
```
to not conflict with translation if we use it one day here
```suggestion
    def thing(self, **__):
```

std-odoo commented on addons/test_portal/tests/test_mail_links.py:
```diff
@@ -0,0 +1,137 @@
+            'perm_read': True,
+        })
+        with self.subTest("object=portal user=portal[acl+] logged"):
+            obj, url, _ = self.do_flow('test_portal.portal', 'portal', login=True, token=False)
```
same below / above
```suggestion
            obj, url, __ = self.do_flow('test_portal.portal', 'portal', login=True, token=False)
```

std-odoo reviewed (commented):
@xmo-odoo Hi :) 

I just a a few minor comments :) (I don't see any better solution)

xmo-odoo commented on addons/portal/controllers/web.py:
```diff
@@ -21,6 +21,6 @@ def _login_redirect(self, uid, redirect=None):
     @http.route()
     def web_client(self, s_action=None, **kw):
-        if request.session.uid and not request.env['res.users'].sudo().browse(request.session.uid).has_group('base.group_user'):
+        if 'redirect' not in kw and request.session.uid and not request.env['res.users'].sudo().browse(request.session.uid).has_group('base.group_user'):
```
Not sure it really matters as this is used to avoid intercepting baseline behaviour, so worst case scenario this would do something like let the parent method redirect to `request.redirect('/web/login?error=access')`.

Still, `dict.get` does match the behaviour / filter of the parent method so seems good enough

tde-banana-odoo commented:
Side note: all tests can be moved to ``test_mail_full`` which alrrady depends on mail and portal. We now have a test_portal.py file in test_mail_full .

xmo-odoo commented:
@tde-banana-odoo did a new pass, rebased everything, folded `test_portal` into `test_mail_full`.

Apparently 3017392f91418bce8e8725aad21ed0d06e06d754 had kinda sorta partially fixed an overlapping case, but I replaced it entirely since it looks like this PR is more general.

xmo-odoo commented:
> Zboing ?

I'm waiting for any sort of review :shrug:

tde-banana-odoo commented:
You could probably check to launch some testing by Luc (you know, PO stuff) if everything is ok on your side :)

xmo-odoo commented:
> You could probably check to launch some testing by Luc (you know, PO stuff) if everything is ok on your side :)

Check what?
